### DISCUSSION TITLE
Record Plugin: Choose Preferred Mic

### DISCUSSION THREAD

**Discussion by tscritch** [katspaugh/wavesurfer.js#3245] [Q&A] [answered]
It would be great to allow for choosing a mic device when recording.

On initial read through of the Record plugin it looks like it would be straightforward to implement.

I was looking at the `startMic` and `startRecording` functions [here](https://github.com/katspaugh/wavesurfer.js/blob/bac048c/src/plugins/record.ts#L75-L78). An optional param could be added for `deviceId` to be passed to the `getUserMedia` function. 
For example
```typescript
public async startMic(options?: {
  deviceId?: string,
}): Promise<MediaStream> {
...
// pass the deviceId if it exists, otherwise just get the stream as normal
stream = await navigator.mediaDevices.getUserMedia({ audio: deviceId ? { deviceId } : true })
// add error handling in the catch below here (ie if the specific device is unavailable)
...
}
```
[MDN Reference](https://developer.mozilla.org/en-US/docs/Web/API/MediaDevices/getUserMedia#using_getusermedia)
Would love input on choosing "exact" device vs preferring device.

Maybe could even add a helper function to get all the audio device ids as well? 
https://developer.mozilla.org/en-US/docs/Web/API/MediaDevices/enumerateDevices

I'm very willing to contribute 🙋‍♂️

Appreciate all the work put into this package!

**Comment by katspaugh** [comment, marked as answer]
Sounds good, your contribution would be appreciated!

**Comment by tscritch** [comment]
Great! I'll get started soon! Do you mind if I add on to the existing Record example?
